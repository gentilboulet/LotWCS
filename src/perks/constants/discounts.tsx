export const DISCOUNT_CHI = "DISCOUNT_CHI";
export type DISCOUNT_CHI = typeof DISCOUNT_CHI;

export const DISCOUNT_SKILL = "DISCOUNT_SKILL";
export type DISCOUNT_SKILL = typeof DISCOUNT_SKILL;

export const DISCOUNT_LORESHEET = "DISCOUNT_LORESHEET";
export type DISCOUNT_LORESHEET = typeof DISCOUNT_LORESHEET;

export const DISCOUNT_LORESHEET_OPTION = "DISCOUNT_LORESHEET_OPTION";
export type DISCOUNT_LORESHEET_OPTION = typeof DISCOUNT_LORESHEET_OPTION;

export const DISCOUNT_KUNGFU_STYLE = "DISCOUNT_KUNGFU_STYLE";
export type DISCOUNT_KUNGFU_STYLE = typeof DISCOUNT_KUNGFU_STYLE;

export const DISCOUNT_KUNGFU_TECHNIQUE = "DISCOUNT_KUNGFU_TECHNIQUE";
export type DISCOUNT_KUNGFU_TECHNIQUE = typeof DISCOUNT_KUNGFU_TECHNIQUE;

export const DISCOUNT_VIRTUE = "DISCOUNT_VIRTUE";
export type DISCOUNT_VIRTUE = typeof DISCOUNT_VIRTUE;
