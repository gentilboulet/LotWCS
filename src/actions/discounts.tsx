import * as constants from '../constants/discounts';
import { IDiscountSkill } from '../types/discounts';
import { IDataSkill } from '../types/skills';
import { skills } from '../data/skills';

export function skill(v: number, subset?: string[]): IDiscountSkill {
  if (subset) {
    return { type: constants.DISCOUNT_SKILL, value: v, skills: subset};
  } else {
    return {
      type: constants.DISCOUNT_SKILL,
      value: v, 
      skills: skills.map((s: IDataSkill) => { return s.name; })
    };
  }
}
