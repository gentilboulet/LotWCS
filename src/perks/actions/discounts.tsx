import * as constants from 'perks/constants/discounts';
import { IDiscountSkill } from 'perks/types/discounts';

import * as dataSkills from 'data/skills';

export function discountSkillFactory(value: number, subset?: string[]): IDiscountSkill {
  if (subset && subset.length > 0) {
    subset.map((skill: string) => dataSkills.validateSkill(skill));
    return { skills: subset, type: constants.DISCOUNT_SKILL, value, };
  } else {
    return {
      skills: dataSkills.skills.map((s: dataSkills.IDataSkill) => s.name),
      type: constants.DISCOUNT_SKILL,
      value,
    };
  }
}
