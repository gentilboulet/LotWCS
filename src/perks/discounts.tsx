import { TChiName } from "../data/chi";
import { KUNGFU_TYPE } from "../data/kungfu/types";
import { TSkillName } from "../data/skills";
import * as constants from "./constants/discounts";

export interface IDiscountSkill {
  type: constants.DISCOUNT_SKILL;
  skills: TSkillName[];
  value: number;
}

export interface IDiscountChi {
  type: constants.DISCOUNT_CHI;
  chis: TChiName[];
  value: number;
}

export interface IDiscountLoresheet {
  type: constants.DISCOUNT_LORESHEET;
  uids: string[];
  value: number;
}

export interface IDiscountLoresheetOption {
  type: constants.DISCOUNT_LORESHEET_OPTION;
  uids: Array<{ lsUid: string; optUid: string[] }>;
  value: number;
}

export interface IDiscountKungfu {
  type: constants.DISCOUNT_KUNGFU_STYLE;
  uids: string[];
  kfType: KUNGFU_TYPE[];
  value: number;
}

export interface IDiscountKungfuTechnique {
  type: constants.DISCOUNT_KUNGFU_TECHNIQUE;
  uids: Array<{ styleUid: string }>;
  kfType: KUNGFU_TYPE[];
  value: number;
}

export type IDiscount =
  | IDiscountSkill
  | IDiscountChi
  | IDiscountLoresheet
  | IDiscountLoresheetOption
  | IDiscountKungfu
  | IDiscountKungfuTechnique;

export function isDiscount(r: any): boolean {
  switch (r.type) {
    case constants.DISCOUNT_CHI:
    case constants.DISCOUNT_SKILL:
    case constants.DISCOUNT_LORESHEET:
    case constants.DISCOUNT_LORESHEET_OPTION:
    case constants.DISCOUNT_KUNGFU_STYLE:
    case constants.DISCOUNT_KUNGFU_TECHNIQUE:
      return true;
    default:
      return false;
  }
}

export type TDiscountsState = IDiscount[];

export function createState(): TDiscountsState {
  return [];
}
