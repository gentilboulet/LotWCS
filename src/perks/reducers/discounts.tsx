import * as Immutable from 'immutable';

import { ICost } from 'costs/types';
import { IDiscount, isDiscount } from 'perks/types/discounts';
import { IStoreState } from 'state/types';

export function pushToDiscounts(oldState: IStoreState, discounts: IDiscount[]): IStoreState {
  return oldState.withMutations((state: IStoreState) => {
    discounts
      .filter((r: IDiscount) => isDiscount(r) )
      .forEach((r: IDiscount) => { state.updateIn(['discounts'], (list) => list.push(r) ); });
  });
}

export function updateDiscounts(oldState: IStoreState, cost: ICost) {
  if (cost.discountIdx < 0) { return oldState; }
  return oldState.withMutations((state: IStoreState) => {
    if (cost.discountNewValue === 0) {
      // Remove the discount
      state.updateIn(['discounts'], (list: Immutable.List<IDiscount>) => list.delete(cost.discountIdx));
    } else {
      // Update the discount
      state.updateIn(['discounts', cost.discountIdx], (r: IDiscount) => {
        r.value = cost.discountNewValue;
        return r;
      });
    }
  });
}
