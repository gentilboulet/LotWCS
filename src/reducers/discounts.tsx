import * as Immutable from 'immutable';
import { IDiscount, isDiscount } from '../types/discounts';
import { ICost } from '../types/costs';
import { IStoreState } from '../types/state';

export function pushToDiscounts(oldState: IStoreState, discounts: IDiscount[]): IStoreState {
  return oldState.withMutations((state: IStoreState) => {
    discounts
      .filter((r: IDiscount) => { return isDiscount(r); })
      .forEach((r: IDiscount) => {
          state.updateIn(['discounts'], (list) => { return list.push(r); });
      });
  });
}

export function updateDiscounts(oldState: IStoreState, cost: ICost) {
  if (cost.discountIdx < 0) { return oldState; }
  return oldState.withMutations((state: IStoreState) => {
    if (cost.discountNewValue === 0) {
      // Remove the discount
      state.updateIn(['discounts'], (list: Immutable.List<IDiscount>) => {
        return list.delete(cost.discountIdx);
      });
    } else {
      // Update the discount
      state.updateIn(['discounts', cost.discountIdx], (r: IDiscount) => {
        r.value = cost.discountNewValue;
        return r;
      });
    }
  });
}
