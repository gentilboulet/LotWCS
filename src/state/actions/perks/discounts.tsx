import * as constants from 'state/constants/perks/discounts';
import { IDiscountSkill } from 'state/discounts';

import { skills, TSkillName } from 'data/skills';

export function discountSkillFactory(value: number, subset?: TSkillName[]): IDiscountSkill {
  if (subset && subset.length > 0) {
    return { skills: subset, type: constants.DISCOUNT_SKILL, value, };
  } else {
    return {
      skills: Object.keys(skills) as TSkillName[],
      type: constants.DISCOUNT_SKILL,
      value,
    };
  }
}
