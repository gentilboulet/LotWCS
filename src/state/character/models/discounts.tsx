import { ICost } from "./costs";
import { IStoreState } from "./type";

import { IDiscount, isDiscount } from "../perks/discounts";

export type TDiscountsState = IDiscount[];

export function createState(): TDiscountsState {
  return [];
}

export function pushToDiscounts(
  state: IStoreState,
  discounts: IDiscount[]
): void {
  discounts
    .filter((r: IDiscount) => isDiscount(r))
    .forEach((r: IDiscount) => {
      state.discounts.push(r);
    });
}

export function getDiscountIndexes(
  state: IStoreState,
  predicate: (d: IDiscount) => boolean
) {
  const discountIdx = state.discounts
    .map((discount: IDiscount, idx: number) => {
      if (predicate(discount)) {
        return idx;
      }
      return NaN;
    })
    .filter((value: number) => !isNaN(value));
  if (discountIdx) {
    return discountIdx;
  } else {
    return [];
  }
}

export function updateDiscounts(state: IStoreState, cost: ICost): void {
  // Update the discount
  if (!cost.discounts) {
    return;
  }

  cost.discounts.forEach(costDiscount => {
    state.discounts[costDiscount.idx].value = costDiscount.newValue;

    if (costDiscount.newValue === 0) {
      // Remove the empty discount
      state.discounts.splice(costDiscount.idx, 1);
    }
  });
}
