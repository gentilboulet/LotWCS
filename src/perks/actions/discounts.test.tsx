import * as discounts from './discounts';

describe('Testing discount creators', () => {
  it('should create a skill discount', () => {
    const discount = discounts.discountSkillFactory(10);
    expect( discount ).toMatchSnapshot();
  });

  it('should create a skill discount for a single skill', () => {
    const discount = discounts.discountSkillFactory(10, ['Awareness', 'Politics']);
    expect( discount ).toMatchSnapshot();
  });

  it('should create a skill discount with an empty subset', () => {
    const discount = discounts.discountSkillFactory(10, []);
    expect( discount ).toMatchSnapshot();
  });
});
