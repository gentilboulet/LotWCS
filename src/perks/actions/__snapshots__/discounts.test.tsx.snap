// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing discount creators should create a skill discount 1`] = `
Object {
  "skills": Array [
    "Awareness",
    "Confidence",
    "Crafting",
    "Finesse",
    "Hardiness",
    "Inspire",
    "Learning",
    "Medicine",
    "Might",
    "Perform",
    "Politics",
    "Ride",
    "Stealth",
    "Survival",
    "Tactics",
    "Wu Wei",
  ],
  "type": "DISCOUNT_SKILL",
  "value": 10,
}
`;

exports[`Testing discount creators should create a skill discount for a single skill 1`] = `
Object {
  "skills": Array [
    "Awareness",
    "Politics",
  ],
  "type": "DISCOUNT_SKILL",
  "value": 10,
}
`;

exports[`Testing discount creators should create a skill discount with an empty subset 1`] = `
Object {
  "skills": Array [
    "Awareness",
    "Confidence",
    "Crafting",
    "Finesse",
    "Hardiness",
    "Inspire",
    "Learning",
    "Medicine",
    "Might",
    "Perform",
    "Politics",
    "Ride",
    "Stealth",
    "Survival",
    "Tactics",
    "Wu Wei",
  ],
  "type": "DISCOUNT_SKILL",
  "value": 10,
}
`;
